mypaint-brush.c

```diff
@@ -719,39 +764,157 @@ smallest_angular_difference(float a, float b)
+      self->states[MYPAINT_BRUSH_STATE_SMUDGE_RA] = mix_colors((self->states[MYPAINT_BRUSH_STATE_SMUDGE_RA]), (r*a), fac, self->settings_value[MYPAINT_BRUSH_SETTING_SMUDGE_LINEAR_MODEL], self->settings_value[MYPAINT_BRUSH_SETTING_SMUDGE_ADD_SUB]);
+      self->states[MYPAINT_BRUSH_STATE_SMUDGE_GA] = mix_colors((self->states[MYPAINT_BRUSH_STATE_SMUDGE_GA]), (g*a), fac, self->settings_value[MYPAINT_BRUSH_SETTING_SMUDGE_LINEAR_MODEL], self->settings_value[MYPAINT_BRUSH_SETTING_SMUDGE_ADD_SUB]);
+      self->states[MYPAINT_BRUSH_STATE_SMUDGE_BA] = mix_colors((self->states[MYPAINT_BRUSH_STATE_SMUDGE_BA]), (b*a), fac, self->settings_value[MYPAINT_BRUSH_SETTING_SMUDGE_LINEAR_MODEL], self->settings_value[MYPAINT_BRUSH_SETTING_SMUDGE_ADD_SUB]);
     }
 
     // color part
-
     float color_h = mypaint_mapping_get_base_value(self->settings[MYPAINT_BRUSH_SETTING_COLOR_H]);
     float color_s = mypaint_mapping_get_base_value(self->settings[MYPAINT_BRUSH_SETTING_COLOR_S]);
     float color_v = mypaint_mapping_get_base_value(self->settings[MYPAINT_BRUSH_SETTING_COLOR_V]);
     float eraser_target_alpha = 1.0;
+    float brush_h, brush_s, brush_l, smudge_h, smudge_s, smudge_l;
+
     if (self->settings_value[MYPAINT_BRUSH_SETTING_SMUDGE] > 0.0) {
-      // mix (in RGB) the smudge color with the brush color
-      hsv_to_rgb_float (&color_h, &color_s, &color_v);
       float fac = self->settings_value[MYPAINT_BRUSH_SETTING_SMUDGE];
+      hsv_to_rgb_float (&color_h, &color_s, &color_v);
+      
       if (fac > 1.0) fac = 1.0;
-      // If the smudge color somewhat transparent, then the resulting
-      // dab will do erasing towards that transparency level.
-      // see also ../doc/smudge_math.png
-      eraser_target_alpha = (1-fac)*1.0 + fac*self->states[MYPAINT_BRUSH_STATE_SMUDGE_A];
-      // fix rounding errors (they really seem to happen in the previous line)
-      eraser_target_alpha = CLAMP(eraser_target_alpha, 0.0, 1.0);
+        // If the smudge color somewhat transparent, then the resulting
+        // dab will do erasing towards that transparency level.
+        // see also ../doc/smudge_math.png
+        eraser_target_alpha = (1-fac)*1.0 + fac*self->states[MYPAINT_BRUSH_STATE_SMUDGE_A];
+        // fix rounding errors (they really seem to happen in the previous line)
+        eraser_target_alpha = CLAMP(eraser_target_alpha, 0.0, 1.0);
+        if (eraser_target_alpha > 0) {
+          //store values for brush and smudge for later adjustments via the smudge adjustment mode setting
+          //using HSL naming but it will depend on the mode selected
+          
+          brush_h = color_h;
+          brush_s = color_s;
+          brush_l = color_v;
+          
+          //grab the RGB w/o alpha multiplied already- we're only using it for colorfulness and brightness later
+          smudge_h = self->states[MYPAINT_BRUSH_STATE_LAST_GETCOLOR_R];
```

Comment:
oops, this is probably wrong.  I should be getting the smudge color, not the LAST_GETCOLOR.  I guess I could divide smudge color it by smudge_alpha to get the color I want here.